// wikijs/dist/wiki.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.default = wiki;

require('babel-polyfill');

var _underscore = require('underscore');

var _underscore2 = _interopRequireDefault(_underscore);

var _util = require('./util');

var _page = require('./page');

var _page2 = _interopRequireDefault(_page);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @namespace
 * @constant
 * @property {string} apiUrl - URL of Wikipedia API
 */
var defaultOptions = {
	apiUrl: 'http://en.wikipedia.org/w/api.php'
};

/**
 * wiki
 * @example
 * wiki({ apiUrl: 'http://fr.wikipedia.org/w/api.php' }).search(...);
 * @namespace Wiki
 * @param  {Object} options
 * @return {Object} - wiki (for chaining methods)
*/
function wiki() {
	var options = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];


	if (this instanceof wiki) {
		console.log('Please do not use wikijs ^1.0.0 as a class. Please see the new README.'); // eslint-disable-line
	}

	var apiOptions = Object.assign({}, defaultOptions, options);

	function handleRedirect(res) {
		if (res.query.redirects && res.query.redirects.length === 1) {
			return (0, _util.api)(apiOptions, {
				prop: 'info|pageprops',
				inprop: 'url',
				ppprop: 'disambiguation',
				titles: res.query.redirects[0].to
			});
		}
		return res;
	}

	/**
  * Search articles
  * @example
  * wiki.search('star wars').then(data => console.log(data.results.length));
  * @example
  * wiki.search('star wars').then(data => {
  * 	data.next().then(...);
  * });
  * @method Wiki#search
  * @param  {string} query - keyword query
  * @param  {Number} [limit] - limits the number of results
  * @return {Promise} - pagination promise with results and next page function
  */
	function search(query) {
		var limit = arguments.length <= 1 || arguments[1] === undefined ? 50 : arguments[1];

		return (0, _util.pagination)(apiOptions, {
			list: 'search',
			srsearch: query,
			srlimit: limit
		}, function (res) {
			return _underscore2.default.pluck(res.query.search, 'title');
		});
	}

	/**
  * Random articles
  * @example
  * wiki.random(3).then(results => console.log(results[0]));
  * @method Wiki#random
  * @param  {Number} [limit] - limits the number of random articles
  * @return {Promise} - List of page titles
  */
	function random() {
		var limit = arguments.length <= 0 || arguments[0] === undefined ? 1 : arguments[0];

		return (0, _util.api)(apiOptions, {
			list: 'random',
			rnnamespace: 0,
			rnlimit: limit
		}).then(function (res) {
			return _underscore2.default.pluck(res.query.random, 'title');
		});
	}

	/**
  * Get Page
  * @example
  * wiki.page('Batman').then(page => console.log(page.pageid));
  * @method Wiki#page
  * @param  {string} title - title of article
  * @return {Promise}
  */
	function page(title) {
		return (0, _util.api)(apiOptions, {
			prop: 'info|pageprops',
			inprop: 'url',
			ppprop: 'disambiguation',
			titles: title
		}).then(handleRedirect).then(function (res) {
			var id = Object.keys(res.query.pages)[0];
			if (!id) {
				throw new Error('No article found');
			}
			return (0, _page2.default)(res.query.pages[id], apiOptions);
		});
	}

	/**
  * Geographical Search
  * @example
  * wiki.geoSearch(32.329, -96.136).then(titles => console.log(titles.length));
  * @method Wiki#geoSearch
  * @param  {Number} lat - latitude
  * @param  {Number} lon - longitude
  * @param  {Number} [radius=1000] - search radius in kilometers (default: 1km)
  * @return {Promise} - List of page titles
  */
	function geoSearch(lat, lon) {
		var radius = arguments.length <= 2 || arguments[2] === undefined ? 1000 : arguments[2];

		return (0, _util.api)(apiOptions, {
			list: 'geosearch',
			gsradius: radius,
			gscoord: lat + '|' + lon
		}).then(function (res) {
			return _underscore2.default.pluck(res.query.geosearch, 'title');
		});
	}

	return {
		search: search,
		random: random,
		page: page,
		geoSearch: geoSearch,
		options: options
	};
}
